File: packages/core/src/research_engine/domain/events.py
"""Event and timeline domain types."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from research_engine.domain.common import DatePrecision


class FuzzyDate(BaseModel):
    """A date with explicit precision."""

    start: datetime
    end: datetime
    precision: DatePrecision


class Event(BaseModel):
    """An event in the timeline."""

    id: UUID
    event_type: str
    timestamp_start: datetime | None = None
    timestamp_end: datetime | None = None
    precision: DatePrecision | None = None
    location_id: UUID | None = None
    location_text: str | None = None
    source_passage_id: UUID | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    confidence: float = 1.0
    created_at: datetime


class EventDraft(BaseModel):
    """Data needed to create an event."""

    event_type: str
    timestamp_start: datetime | None = None
    timestamp_end: datetime | None = None
    precision: DatePrecision | None = None
    location_id: UUID | None = None
    location_text: str | None = None
    source_passage_id: UUID | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    confidence: float = 1.0


class EventActor(BaseModel):
    """An entity participating in an event."""

    event_id: UUID
    entity_id: UUID
    role: str


class EventFilter(BaseModel):
    """Filters for querying events."""

    event_types: list[str] | None = None
    actor_entity_ids: list[UUID] | None = None
    date_range_start: datetime | None = None
    date_range_end: datetime | None = None
    location_id: UUID | None = None
    payload: dict[str, Any] | None = None


class TimelineBucket(BaseModel):
    """A time bucket in a timeline aggregation."""

    bucket: str
    count: int
    aggregates: dict[str, Any] = Field(default_factory=dict)


class TimelineStream(BaseModel):
    """A named stream in a timeline comparison."""

    name: str
    buckets: list[TimelineBucket]

File: crates/marginalia-types/src/events.rs
//! Event and timeline types mirroring `domain/events.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::common::DatePrecision;

/// A date with explicit precision.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FuzzyDate {
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
    pub precision: DatePrecision,
}

/// An event in the timeline.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Event {
    pub id: Uuid,
    pub event_type: String,
    #[serde(default)]
    pub timestamp_start: Option<DateTime<Utc>>,
    #[serde(default)]
    pub timestamp_end: Option<DateTime<Utc>>,
    #[serde(default)]
    pub precision: Option<DatePrecision>,
    #[serde(default)]
    pub location_id: Option<Uuid>,
    #[serde(default)]
    pub location_text: Option<String>,
    #[serde(default)]
    pub source_passage_id: Option<Uuid>,
    #[serde(default)]
    pub payload: Map<String, Value>,
    #[serde(default = "default_confidence")]
    pub confidence: f64,
    pub created_at: DateTime<Utc>,
}

fn default_confidence() -> f64 {
    1.0
}

/// Data needed to create an event.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EventDraft {
    pub event_type: String,
    #[serde(default)]
    pub timestamp_start: Option<DateTime<Utc>>,
    #[serde(default)]
    pub timestamp_end: Option<DateTime<Utc>>,
    #[serde(default)]
    pub precision: Option<DatePrecision>,
    #[serde(default)]
    pub location_id: Option<Uuid>,
    #[serde(default)]
    pub location_text: Option<String>,
    #[serde(default)]
    pub source_passage_id: Option<Uuid>,
    #[serde(default)]
    pub payload: Map<String, Value>,
    #[serde(default = "default_confidence")]
    pub confidence: f64,
}

/// An entity participating in an event.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EventActor {
    pub event_id: Uuid,
    pub entity_id: Uuid,
    pub role: String,
}

/// Filters for querying events.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct EventFilter {
    #[serde(default)]
    pub event_types: Option<Vec<String>>,
    #[serde(default)]
    pub actor_entity_ids: Option<Vec<Uuid>>,
    #[serde(default)]
    pub date_range_start: Option<DateTime<Utc>>,
    #[serde(default)]
    pub date_range_end: Option<DateTime<Utc>>,
    #[serde(default)]
    pub location_id: Option<Uuid>,
    #[serde(default)]
    pub payload: Option<Map<String, Value>>,
}

/// A time bucket in a timeline aggregation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TimelineBucket {
    pub bucket: String,
    pub count: i64,
    #[serde(default)]
    pub aggregates: Map<String, Value>,
}

/// A named stream in a timeline comparison.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TimelineStream {
    pub name: String,
    pub buckets: Vec<TimelineBucket>,
}
